File: gateway/src/test/java/com/beat/gateway/GatewayConfigGroupTest.java
package com.beat.gateway;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

import org.junit.jupiter.api.Test;

class GatewayConfigGroupTest {

	@Test
	void gatewayConfigGroupsExposeOptionalStatefulSecurityServices() {
		assertArrayEquals(
			new GatewayConfigGroup[] {
				GatewayConfigGroup.REFRESH_TOKEN_STORE,
				GatewayConfigGroup.GUEST_ACCESS
			},
			GatewayConfigGroup.values()
		);
	}

	@Test
	void guestAccessGroupOwnsGuestSessionPasswordAndThrottleServices() throws IOException {
		assertEquals(
			"com.beat.gateway.guest.internal.config.GuestAccessConfig",
			GatewayConfigGroup.GUEST_ACCESS.getConfigClass().getName()
		);

		String source = source("src/main/java/com/beat/gateway/guest/internal/config/GuestAccessConfig.java");

		assertTrue(source.contains("RedisConfig.class"));
		assertTrue(source.contains("GuestSessionService.class"));
		assertTrue(source.contains("GuestPasswordHashService.class"));
		assertTrue(source.contains("GuestAccessThrottleService.class"));
	}

	@Test
	void publicServletSecurityAnnotationIsStaticImportSurfaceForExecutableModules() throws IOException {
		String annotationSource = source("src/main/java/com/beat/gateway/EnableGatewayServletSecurity.java");
		String source = source("src/main/java/com/beat/gateway/authentication/internal/config/ServletSecurityConfig.java");

		assertTrue(annotationSource.contains("@Import(ServletSecurityConfig.class)"));
		assertTrue(source.contains("JwtConfig.class"));
		assertTrue(source.contains("SecurityFilterConfig.class"));
		assertTrue(source.contains("WebMvcConfig.class"));
		assertFalse(source.contains("RefreshTokenConfig.class"));
		assertFalse(source.contains("RedisConfig.class"));
		assertFalse(source.contains("RefreshTokenService.class"));
	}

	@Test
	void securityMdcFilterIsPartOfServletSecurityGroupAndAutoRegistrationIsDisabled() throws IOException {
		String servletSource =
			source("src/main/java/com/beat/gateway/authentication/internal/config/SecurityFilterConfig.java");

		assertTrue(servletSource.contains("gatewaySecurityMdcLoggingFilter"));
		assertTrue(servletSource.contains("FilterRegistrationBean<SecurityMdcLoggingFilter>"));
		assertTrue(servletSource.contains("registration.setEnabled(false)"));
	}

	@Test
	void refreshTokenStoreGroupOwnsRedisRepositoryAndRefreshTokenService() throws IOException {
		assertEquals(
			"com.beat.gateway.refreshtoken.internal.config.RefreshTokenConfig",
			GatewayConfigGroup.REFRESH_TOKEN_STORE.getConfigClass().getName()
		);

		String source = source("src/main/java/com/beat/gateway/refreshtoken/internal/config/RefreshTokenConfig.java");

		assertTrue(source.contains("RedisConfig.class"));
		assertTrue(source.contains("RefreshTokenService.class"));
	}

	@Test
	void jwtFilterEnrichesAlreadyInitializedMdcWithAuthenticatedUser() throws IOException {
		String source = source("src/main/java/com/beat/gateway/authentication/internal/JwtAuthenticationFilter.java");

		assertTrue(source.contains("MDC.put(BaseMdcLoggingFilter.USER_ID_KEY, Long.toString(memberId))"));
	}

	@Test
	void jwtConfigRegistersJwtProviderOnly() throws IOException {
		String source = source("src/main/java/com/beat/gateway/jwt/internal/config/JwtConfig.java");

		assertTrue(source.contains("JwtTokenProvider jwtTokenProvider(JwtProperties jwtProperties)"));
		assertFalse(source.contains("RefreshTokenService.class"));
		assertFalse(source.contains("RedisConfig.class"));
	}

	private String source(String path) throws IOException {
		return Files.readString(Path.of(path));
	}
}

File: gateway/src/test/kotlin/com/beat/gateway/GatewayConfigGroupTest.kt
package com.beat.gateway

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertAll
import java.nio.file.Files
import java.nio.file.Path
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse

/**
 * gateway bootstrap 계약(공개 표면 → internal config 매핑)을 고정한다.
 * config group이 소유하는 bean 집합은 source 검사로 회귀를 막는다.
 */
class GatewayConfigGroupTest {

    @Test
    fun `config group은 optional stateful security 기능만 노출한다`() {
        assertEquals(
            listOf(GatewayConfigGroup.REFRESH_TOKEN_STORE, GatewayConfigGroup.GUEST_ACCESS),
            GatewayConfigGroup.entries,
        )
    }

    @Test
    fun `GUEST_ACCESS group이 guest 세션·비밀번호·스로틀 서비스를 소유한다`() {
        assertEquals(
            "com.beat.gateway.guest.internal.config.GuestAccessConfig",
            GatewayConfigGroup.GUEST_ACCESS.configClass.name,
        )

        val source = source("guest/internal/config/GuestAccessConfig.kt")

        assertAll(
            { assertTrue(source.contains("RedisConfig::class")) },
            { assertTrue(source.contains("GuestSessionService::class")) },
            { assertTrue(source.contains("GuestPasswordHashService::class")) },
            { assertTrue(source.contains("GuestAccessThrottleService::class")) },
        )
    }

    @Test
    fun `servlet security 공개 annotation이 실행 모듈의 static import 표면이다`() {
        val annotationSource = source("EnableGatewayServletSecurity.kt")
        val configSource = source("authentication/internal/config/ServletSecurityConfig.kt")

        assertAll(
            { assertTrue(annotationSource.contains("@Import(ServletSecurityConfig::class)")) },
            { assertTrue(configSource.contains("JwtConfig::class")) },
            { assertTrue(configSource.contains("SecurityFilterConfig::class")) },
            { assertTrue(configSource.contains("WebMvcConfig::class")) },
            { assertFalse(configSource.contains("RefreshTokenConfig::class")) },
            { assertFalse(configSource.contains("RedisConfig::class")) },
            { assertFalse(configSource.contains("RefreshTokenService::class")) },
        )
    }

    @Test
    fun `MDC filter는 servlet security group에 속하고 servlet 자동 등록은 비활성화된다`() {
        val source = source("authentication/internal/config/SecurityFilterConfig.kt")

        assertAll(
            { assertTrue(source.contains("gatewaySecurityMdcLoggingFilter")) },
            { assertTrue(source.contains("FilterRegistrationBean<SecurityMdcLoggingFilter>")) },
            { assertTrue(source.contains("registration.isEnabled = false")) },
        )
    }

    @Test
    fun `REFRESH_TOKEN_STORE group이 Redis repository와 refresh token 서비스를 소유한다`() {
        assertEquals(
            "com.beat.gateway.refreshtoken.internal.config.RefreshTokenConfig",
            GatewayConfigGroup.REFRESH_TOKEN_STORE.configClass.name,
        )

        val source = source("refreshtoken/internal/config/RefreshTokenConfig.kt")

        assertAll(
            { assertTrue(source.contains("RedisConfig::class")) },
            { assertTrue(source.contains("RefreshTokenService::class")) },
        )
    }

    @Test
    fun `JWT filter가 이미 초기화된 MDC에 인증 사용자 id를 채운다`() {
        val source = source("authentication/internal/JwtAuthenticationFilter.kt")

        assertTrue(source.contains("MDC.put(BaseMdcLoggingFilter.USER_ID_KEY, memberId.toString())"))
    }

    @Test
    fun `JwtConfig는 JWT 관련 bean만 등록하고 refresh token·Redis를 가져오지 않는다`() {
        val source = source("jwt/internal/config/JwtConfig.kt")

        assertAll(
            { assertTrue(source.contains("fun jwtTokenProvider(")) },
            { assertTrue(source.contains("JwtTokenProvider")) },
            { assertFalse(source.contains("RefreshTokenService")) },
            { assertFalse(source.contains("RedisConfig")) },
        )
    }

    private fun source(relativePath: String): String =
        Files.readString(Path.of("src/main/kotlin/com/beat/gateway/$relativePath"))
}
